src/program/lwaftr/csv_stats.lua

```diff
@@ -84,41 +71,81 @@ function CSVStatsTimer:add_app(id, links, link_names)
          if name == 'rx' then link = app.input.input end
          if name == 'tx' then link = app.output.output end
       end
-      assert(link, "Link named "..name.." not found in "..id)
-      add_link_data(name, link)
+      if not link then return false end
+      table.insert(resolved_links, {name, link})
    end
+   for _, resolved_link in ipairs(resolved_links) do
+      local name, link = unpack(resolved_link)
+      local link_name = link_names[name] or name
+      local data = {
+         link_name = link_name,
+         txpackets = link.txpackets,
+         txbytes = link.txbytes,
+      }
+      if not self.hydra_mode then
+         local h = (',%s MPPS,%s Gbps'):format(link_name, link_name)
+         self.header = self.header..h
+      end
+      table.insert(self.link_data, data)
+   end
+   return true
+end
+
+-- Add links from an app whose identifier is ID to the CSV timer.  If
+-- present, LINKS is an array of strings identifying a subset of links
+-- to monitor.  The optional LINK_NAMES table maps link names to
+-- human-readable names, for the column headers.
+function CSVStatsTimer:add_app(id, links, link_names)
+   -- Because we are usually measuring counters from another process and
+   -- that process is probably spinning up as we are installing the
+   -- counter, we defer the resolve operation and try to resolve it from
+   -- inside the timer.
+   table.insert(self.deferred_apps, {id, links, link_names})
 end
 
 function CSVStatsTimer:set_period(period) self.period = period end
 
 -- Activate the timer with a period of PERIOD seconds.
-function CSVStatsTimer:activate(write_header)
-   if write_header then
-      self.file:write(self.header..'\n')
-      self.file:flush()
+function CSVStatsTimer:start()
+   local function tick() return self:tick() end
```

Comment:
What's the goal of wrapping the `self:tick()` call instead of calling `self:tick()` directly?

Reply:
It's because timer functions are called without any arguments, and here we need to call tick(self).